src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -371,4 +405,86 @@ isMissingPackets = do
       Nothing -> pure True
       Just ct -> do
         len <- lengthPacketSet pxs
-        pure (fromIntegral len - 1 < ct)
\ No newline at end of file
+        pure (fromIntegral len - 1 < ct)
+
+-- Unsafe ----------------------------------------------------------------------
+
+fromPacketBuildR ::
+  MonadIO m =>
+  Ptr Word8 ->
+  Int ->
+  Packet ByteString ->
+  m (Ptr Word8, Int)
+fromPacketBuildR buffer unused packet = do
+  let build = Encode.reverseMessageBuilder (wireBuildPacket packet)
+  liftIO (fromBuildR build buffer unused)
+
+encodeBufferOffPacket ::
+  MonadIO m =>
+  Ptr Word8 ->
+  Int ->
+  Packet ByteString ->
+  m ByteString
+encodeBufferOffPacket ptr unused packet = do
+  (ptr', unused') <- fromPacketBuildR ptr unused packet
+  size <- liftIO (readTotal ptr' unused')
+  liftIO (unsafePackCStringLen ptr' size)
+
+withEncodeBuffer :: MonadUnliftIO m => Int -> (Ptr Word8 -> m a) -> m a
```

Comment:
It might be a nicer API surface (or it could just be a new function) to have a function like `withEncodePacket :: MonadUnliftIO m => Int -> ((Packet ByteString -> m ByteString) -> m a) -> m a` instead. That way all the pointer shenanigans can stay down here in the unsafe section.

Reply:
I agree this would be nicer but I don't like the idea of performing a CPS transformation on the packet serialization function, especially since it is static.

Reply:
What do you mean by 'static'?

Reply:
By static I mean the function passed in by `((Packet ByteString -> m ByteString) -> m a)` never changes, it will always be `encodePacketOffBuffer`.

Reply:
Right, but it would be closed over the `Ptr` and `bufferSize` so there are a few less inner details that get exposed, and less ways for a user of this function to mess it up.